// bitcore-node/lib/cli/bitcored.js
'use strict';

var Liftoff = require('liftoff');

function main(parentServicesPath, additionalServices) {

  //bitcored.js invokes daemon.js, while bitcore.js invokes main.js

  console.log('===== bitcore-node/lib/cli/bitcored main start...');

  var liftoff = new Liftoff({
    name: 'bitcored',
    moduleName: 'bitcore-node',
    configName: 'bitcore-node',
    processTitle: 'bitcored'
  }).on('require', function (name) {
    console.log('Loading:', name);
  }).on('requireFail', function (name, err) {
    console.log('Unable to load:', name, err);
  }).on('respawn', function (flags, child) {
    console.log('Detected node flags:', flags);
    console.log('Respawned to PID:', child.pid);
  });

  liftoff.launch({
    cwd: process.cwd()
  }, function(env){

    var node;

    if (env.configPath && env.modulePath) {
      console.log('env.configPath '+env.configPath+' env.modulePath '+env.modulePath);
      node = require(env.modulePath);
      node.cli.daemon();
    } else {
      console.log('env.configPath and modulePath is not defined. taking the default setting');
      node = require('../../');

      //bitcore-node/lib/cli/daemon.js (same dir as this file)
      console.log('calling cli.daemon with parentServicesPath '+parentServicesPath+' additionalServices '+additionalServices);
      node.cli.daemon(parentServicesPath, additionalServices);
    }

  });

}

module.exports = main;
